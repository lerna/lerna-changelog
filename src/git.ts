const execa = require("execa");

export function getRootPath() {
  const cwd = process.cwd();
  return execa.sync("git", ["rev-parse", "--show-toplevel"], { cwd }).stdout;
}

export async function changedPaths(sha: string): Promise<string[]> {
  const result = await execa("git", ["show", "-m", "--name-only", "--pretty=format:", "--first-parent", sha]);
  return result.stdout.split("\n");
}

/**
 * All existing tags in the repository
 */
export function listTagNames(): string[] {
  return execa
    .sync("git", ["tag"])
    .stdout.split("\n")
    .filter(Boolean);
}

/**
 * The latest reachable tag starting from HEAD
 */
export function lastTag(): string {
  return execa.sync("git", ["describe", "--abbrev=0", "--tags"]).stdout;
}

export interface CommitListItem {
  sha: string;
  refName: string;
  summary: string;
  date: string;
}

export function parseLogMessage(commit: string): CommitListItem | null {
  const parts = commit.match(/hash<(.+)> ref<(.*)> message<(.*)> date<(.*)>/) || [];

  if (!parts || parts.length === 0) {
    return null;
  }

  return {
    sha: parts[1],
    refName: parts[2],
    summary: parts[3],
    date: parts[4],
  };
}

export function listCommits(from: string, to: string = ""): CommitListItem[] {
  // Prints "hash<short-hash> ref<ref-name> message<summary> date<date>"
  // This format is used in `getCommitInfos` for easily analize the commit.
  return execa
    .sync("git", [
      "log",
      "--oneline",
      "--pretty=hash<%h> ref<%D> message<%s> date<%cd>",
      "--date=short",
      `${from}..${to}`,
    ])
    .stdout.split("\n")
    .filter(Boolean)
    .map(parseLogMessage)
    .filter(Boolean);
}
