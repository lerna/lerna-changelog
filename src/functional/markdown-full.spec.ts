/* tslint:disable:max-line-length */
import { CommitListItem } from "../git";

jest.mock("../../src/progress-bar");
jest.mock("../../src/changelog");
jest.mock("../../src/github-api");
jest.mock("../git");
jest.mock("../fetch");
jest.mock("../packages");

const listOfCommits: CommitListItem[] = [
  {
    sha: "a0000015",
    refName: "",
    summary: "chore: making of episode viii",
    date: "2015-12-18",
  },
  {
    sha: "a0000014",
    refName: "",
    summary: "feat: infiltration (#7)",
    date: "2015-12-18",
  },
  {
    sha: "a0000013",
    refName: "HEAD -> master, tag: v6.0.0, origin/master, origin/HEAD",
    summary: "chore(release): releasing component",
    date: "1983-05-25",
  },
  {
    sha: "a0000012",
    refName: "",
    summary: "Merge pull request #6 from return-of-the-jedi",
    date: "1983-05-25",
  },
  {
    sha: "a0000011",
    refName: "",
    summary: "feat: I am your father (#5)",
    date: "1983-05-25",
  },
  {
    sha: "a0000010",
    refName: "",
    summary: "fix(han-solo): unfreezes (#4)",
    date: "1983-05-25",
  },
  {
    sha: "a0000009",
    refName: "tag: v5.0.0",
    summary: "chore(release): releasing component",
    date: "1980-05-17",
  },
  {
    sha: "a0000008",
    refName: "",
    summary: "Merge pull request #3 from empire-strikes-back",
    date: "1980-05-17",
  },
  {
    sha: "a0000007",
    refName: "",
    summary: "fix: destroy rebels base",
    date: "1980-05-17",
  },
  {
    sha: "a0000006",
    refName: "",
    summary: "chore: the end of Alderaan (#2)",
    date: "1980-05-17",
  },
  {
    sha: "a0000005",
    refName: "",
    summary: "refactor(death-star): add deflector shield",
    date: "1980-05-17",
  },
  {
    sha: "a0000004",
    refName: "tag: v4.0.0",
    summary: "chore(release): releasing component",
    date: "1977-05-25",
  },
  {
    sha: "a0000003",
    refName: "",
    summary: "Merge pull request #1 from star-wars",
    date: "1977-05-25",
  },
  {
    sha: "a0000002",
    refName: "tag: v0.1.0",
    summary: "chore(release): releasing component",
    date: "1966-01-01",
  },
  {
    sha: "a0000001",
    refName: "",
    summary: "fix: some random fix which will be ignored",
    date: "1966-01-01",
  },
];

const listOfTags = ["v6.0.0", "v5.0.0", "v4.0.0", "v3.0.0", "v2.0.0", "v1.0.0", "v0.1.0"];

const listOfPackagesForEachCommit: { [id: string]: string[] } = {
  a0000001: ["packages/random/foo.js"],
  a0000002: ["packages/random/package.json"],
  a0000003: ["packages/star-wars/a-new-hope/rebels.js"],
  a0000004: ["packages/star-wars/a-new-hope/package.json"],
  a0000005: ["packages/star-wars/empire-strikes-back/death-star.js"],
  a0000006: ["packages/star-wars/empire-strikes-back/death-star.js"],
  a0000007: ["packages/star-wars/empire-strikes-back/hoth.js"],
  a0000008: ["packages/star-wars/empire-strikes-back/hoth.js"],
  a0000009: ["packages/star-wars/empire-strikes-back/package.json"],
  a0000010: ["packages/star-wars/return-of-the-jedi/jabba-the-hutt.js"],
  a0000011: ["packages/star-wars/return-of-the-jedi/vader-luke.js"],
  a0000012: ["packages/star-wars/return-of-the-jedi/leia.js"],
  a0000013: ["packages/star-wars/return-of-the-jedi/package.json"],
  a0000014: [
    "packages/star-wars/the-force-awakens/mission.js",
    "packages/star-wars/origin-stories/rogue-one/mission.js",
  ],
  a0000015: ["packages/untitled/script.md"],
};

const listOfFileForEachCommit: { [id: string]: string[] } = {
  a0000001: ["random/foo.js"],
  a0000002: ["random/package.json"],
  a0000003: ["a-new-hope/rebels.js"],
  a0000004: ["a-new-hope/package.json"],
  a0000005: ["empire-strikes-back/death-star.js"],
  a0000006: ["empire-strikes-back/death-star.js"],
  a0000007: ["empire-strikes-back/hoth.js"],
  a0000008: ["empire-strikes-back/hoth.js"],
  a0000009: ["empire-strikes-back/package.json"],
  a0000010: ["return-of-the-jedi/jabba-the-hutt.js"],
  a0000011: ["return-of-the-jedi/vader-luke.js"],
  a0000012: ["return-of-the-jedi/leia.js"],
  a0000013: ["return-of-the-jedi/package.json"],
  a0000014: ["the-force-awakens/mission.js", "rogue-one/mission.js"],
  a0000015: ["untitled/script.md"],
};

const usersCache = {
  "https://api.github.com/users/luke": {
    body: {
      login: "luke",
      html_url: "https://github.com/luke",
      name: "Luke Skywalker",
    },
  },
  "https://api.github.com/users/princess-leia": {
    body: {
      login: "princess-leia",
      html_url: "https://github.com/princess-leia",
      name: "Princess Leia Organa",
    },
  },
  "https://api.github.com/users/vader": {
    body: {
      login: "vader",
      html_url: "https://github.com/vader",
      name: "Darth Vader",
    },
  },
  "https://api.github.com/users/gtarkin": {
    body: {
      login: "gtarkin",
      html_url: "https://github.com/gtarkin",
      name: "Governor Tarkin",
    },
  },
  "https://api.github.com/users/han-solo": {
    body: {
      login: "han-solo",
      html_url: "https://github.com/han-solo",
      name: "Han Solo",
    },
  },
  "https://api.github.com/users/chewbacca": {
    body: {
      login: "chewbacca",
      html_url: "https://github.com/chewbacca",
      name: "Chwebacca",
    },
  },
  "https://api.github.com/users/rd-d2": {
    body: {
      login: "rd-d2",
      html_url: "https://github.com/rd-d2",
      name: "R2-D2",
    },
  },
  "https://api.github.com/users/c-3po": {
    body: {
      login: "c-3po",
      html_url: "https://github.com/c-3po",
      name: "C-3PO",
    },
  },
};
const issuesCache = {
  "https://api.github.com/repos/lerna/lerna-changelog/issues/1": {
    body: {
      number: 1,
      title: "feat: May the force be with you",
      labels: [{ name: "Type: New Feature" }],
      pull_request: {
        html_url: "https://github.com/lerna/lerna-changelog/pull/1",
      },
      user: usersCache["https://api.github.com/users/luke"].body,
    },
  },
  "https://api.github.com/repos/lerna/lerna-changelog/issues/2": {
    body: {
      number: 2,
      title: "chore: Terminate her... immediately!",
      labels: [{ name: "Type: Breaking Change" }],
      pull_request: {
        html_url: "https://github.com/lerna/lerna-changelog/pull/2",
      },
      user: usersCache["https://api.github.com/users/gtarkin"].body,
    },
  },
  "https://api.github.com/repos/lerna/lerna-changelog/issues/3": {
    body: {
      number: 3,
      title: "fix: Get me the rebels base!",
      labels: [{ name: "Type: Bug" }],
      pull_request: {
        html_url: "https://github.com/lerna/lerna-changelog/pull/3",
      },
      user: usersCache["https://api.github.com/users/vader"].body,
    },
  },
  "https://api.github.com/repos/lerna/lerna-changelog/issues/4": {
    body: {
      number: 4,
      title: "fix: RRRAARRWHHGWWR",
      labels: [{ name: "Type: Bug" }, { name: "Type: Maintenance" }],
      pull_request: {
        html_url: "https://github.com/lerna/lerna-changelog/pull/4",
      },
      user: usersCache["https://api.github.com/users/chewbacca"].body,
    },
  },
  "https://api.github.com/repos/lerna/lerna-changelog/issues/5": {
    body: {
      number: 5,
      title: "feat: I am your father",
      labels: [{ name: "Type: New Feature" }],
      pull_request: {
        html_url: "https://github.com/lerna/lerna-changelog/pull/5",
      },
      user: usersCache["https://api.github.com/users/vader"].body,
    },
  },
  "https://api.github.com/repos/lerna/lerna-changelog/issues/6": {
    body: {
      number: 6,
      title: "refactor: he is my brother",
      labels: [{ name: "Type: Enhancement" }],
      pull_request: {
        html_url: "https://github.com/lerna/lerna-changelog/pull/6",
      },
      user: usersCache["https://api.github.com/users/princess-leia"].body,
    },
  },
  "https://api.github.com/repos/lerna/lerna-changelog/issues/7": {
    body: {
      number: 7,
      title: "feat: that is not how the Force works!",
      labels: [{ name: "Type: New Feature" }, { name: "Type: Enhancement" }],
      pull_request: {
        html_url: "https://github.com/lerna/lerna-changelog/pull/7",
      },
      user: usersCache["https://api.github.com/users/han-solo"].body,
    },
  },
};
const listOfWorkspacePackages = [
  {
    name: "random",
    location: "/path/to/project/packages/random",
  },
  {
    name: "@star-wars/a-new-hope",
    location: "/path/to/project/packages/star-wars/a-new-hope",
  },
  {
    name: "@star-wars/empire-strikes-back",
    location: "/path/to/project/packages/star-wars/empire-strikes-back",
  },
  {
    name: "@star-wars/return-of-the-jedi",
    location: "/path/to/project/packages/star-wars/return-of-the-jedi",
  },
  {
    name: "@star-wars/the-force-awakens",
    location: "/path/to/project/packages/star-wars/the-force-awakens",
  },
  {
    name: "@star-wars/rogue-one",
    location: "/path/to/project/packages/star-wars/origin-stories/rogue-one",
  },
  {
    name: "@star-wars/solo",
    location: "/path/to/project/packages/star-wars/origin-stories/solo",
  },
];

describe("createMarkdown", () => {
  beforeEach(() => {
    require("../fetch").__resetMockResponses();
    require("../packages").getPackages.mockImplementation(() => listOfWorkspacePackages);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe("single tags", () => {
    it("outputs correct changelog", async () => {
      require("../git").changedPaths.mockImplementation((sha: string) => listOfPackagesForEachCommit[sha]);
      require("../git").lastTag.mockImplementation(() => "v8.0.0");
      require("../git").listCommits.mockImplementation(() => listOfCommits);
      require("../git").listTagNames.mockImplementation(() => listOfTags);
      require("../fetch").__setMockResponses({
        ...usersCache,
        ...issuesCache,
      });

      const MockedChangelog = require("../changelog").default;
      const changelog = new MockedChangelog();

      const markdown = await changelog.createMarkdown();

      expect(markdown).toMatchSnapshot();
    });
  });

  describe("multiple tags", () => {
    it("outputs correct changelog", async () => {
      require("../git").changedPaths.mockImplementation((sha: string) => listOfPackagesForEachCommit[sha]);
      require("../git").lastTag.mockImplementation(() => "v8.0.0");
      require("../git").listCommits.mockImplementation(() => [
        {
          sha: "a0000004",
          refName: "tag: a-new-hope@4.0.0, tag: empire-strikes-back@5.0.0, tag: return-of-the-jedi@6.0.0",
          summary: "chore(release): releasing component",
          date: "1977-05-25",
        },
        {
          sha: "a0000003",
          refName: "",
          summary: "Merge pull request #1 from star-wars",
          date: "1977-05-25",
        },
        {
          sha: "a0000002",
          refName: "tag: v0.1.0",
          summary: "chore(release): releasing component",
          date: "1966-01-01",
        },
        {
          sha: "a0000001",
          refName: "",
          summary: "fix: some random fix which will be ignored",
          date: "1966-01-01",
        },
      ]);
      require("../git").listTagNames.mockImplementation(() => [
        "a-new-hope@4.0.0",
        "attack-of-the-clones@3.1.0",
        "empire-strikes-back@5.0.0",
        "return-of-the-jedi@6.0.0",
        "revenge-of-the-sith@3.0.0",
        "the-force-awakens@7.0.0",
        "the-phantom-menace@1.0.0",
      ]);
      require("../fetch").__setMockResponses({
        ...usersCache,
        ...issuesCache,
      });

      const MockedChangelog = require("../changelog").default;
      const changelog = new MockedChangelog();

      const markdown = await changelog.createMarkdown();

      expect(markdown).toMatchSnapshot();
    });
  });

  describe("single project", () => {
    it("outputs correct changelog", async () => {
      require("../git").changedPaths.mockImplementation((sha: string) => listOfFileForEachCommit[sha]);
      require("../git").lastTag.mockImplementation(() => "v8.0.0");
      require("../git").listCommits.mockImplementation(() => listOfCommits);
      require("../git").listTagNames.mockImplementation(() => listOfTags);
      require("../fetch").__setMockResponses({
        ...usersCache,
        ...issuesCache,
      });

      const MockedChangelog = require("../changelog").default;
      const changelog = new MockedChangelog();

      const markdown = await changelog.createMarkdown();

      expect(markdown).toMatchSnapshot();
    });
  });

  describe("authentication", () => {
    describe("when github token is not valid", () => {
      const badCredentials = {
        message: "Bad credentials",
        documentation_url: "https://developer.github.com/v3",
      };
      beforeEach(() => {
        require("../git").changedPaths.mockImplementation((sha: string) => listOfFileForEachCommit[sha]);
        require("../git").lastTag.mockImplementation(() => "v8.0.0");
        require("../git").listCommits.mockImplementation(() => listOfCommits);
        require("../git").listTagNames.mockImplementation(() => listOfTags);

        const unauthorized = {
          status: 401,
          statusText: "Unauthorized",
          ok: false,
          body: badCredentials,
        };
        require("../fetch").__setMockResponses({
          ...usersCache,
          ...Object.keys(issuesCache).reduce(
            (unauthorizedIssues, issue) => ({
              ...unauthorizedIssues,
              [issue]: unauthorized,
            }),
            {}
          ),
        });
      });
      afterEach(() => {
        jest.resetAllMocks();
      });
      it("should abort with a proper error message", async () => {
        const MockedChangelog = require("../changelog").default;
        const changelog = new MockedChangelog();
        expect.assertions(2);
        try {
          await changelog.createMarkdown();
        } catch (error) {
          expect(error).toEqual(
            expect.objectContaining({ message: expect.stringContaining("Fetch error: Unauthorized") })
          );
          expect(error).toEqual(
            expect.objectContaining({ message: expect.stringContaining(JSON.stringify(badCredentials)) })
          );
        }
      });
    });
  });
});
