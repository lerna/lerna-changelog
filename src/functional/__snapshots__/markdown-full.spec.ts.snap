// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMarkdown multiple tags outputs correct changelog 1`] = `
"
## a-new-hope@4.0.0 (1977-05-25)

#### :rocket: New Feature
* \`@star-wars/a-new-hope\`
  * [#1](https://github.com/lerna/lerna-changelog/pull/1) feat: May the force be with you ([@luke](https://github.com/luke))

#### Committers: 1
- Luke Skywalker ([@luke](https://github.com/luke))


## empire-strikes-back@5.0.0 (1977-05-25)

#### :rocket: New Feature
* \`@star-wars/a-new-hope\`
  * [#1](https://github.com/lerna/lerna-changelog/pull/1) feat: May the force be with you ([@luke](https://github.com/luke))

#### Committers: 1
- Luke Skywalker ([@luke](https://github.com/luke))


## return-of-the-jedi@6.0.0 (1977-05-25)

#### :rocket: New Feature
* \`@star-wars/a-new-hope\`
  * [#1](https://github.com/lerna/lerna-changelog/pull/1) feat: May the force be with you ([@luke](https://github.com/luke))

#### Committers: 1
- Luke Skywalker ([@luke](https://github.com/luke))"
`;

exports[`createMarkdown single project outputs correct changelog 1`] = `
"
## Unreleased (2099-01-01)

#### :rocket: New Feature
* [#7](https://github.com/lerna/lerna-changelog/pull/7) feat: that is not how the Force works! ([@han-solo](https://github.com/han-solo))

#### :nail_care: Enhancement
* [#7](https://github.com/lerna/lerna-changelog/pull/7) feat: that is not how the Force works! ([@han-solo](https://github.com/han-solo))

#### Committers: 1
- Han Solo ([@han-solo](https://github.com/han-solo))


## v6.0.0 (1983-05-25)

#### :rocket: New Feature
* [#5](https://github.com/lerna/lerna-changelog/pull/5) feat: I am your father ([@vader](https://github.com/vader))

#### :bug: Bug Fix
* [#4](https://github.com/lerna/lerna-changelog/pull/4) fix: RRRAARRWHHGWWR ([@chewbacca](https://github.com/chewbacca))

#### :nail_care: Enhancement
* [#6](https://github.com/lerna/lerna-changelog/pull/6) refactor: he is my brother ([@princess-leia](https://github.com/princess-leia))

#### :house: Maintenance
* [#4](https://github.com/lerna/lerna-changelog/pull/4) fix: RRRAARRWHHGWWR ([@chewbacca](https://github.com/chewbacca))

#### Committers: 3
- Chwebacca ([@chewbacca](https://github.com/chewbacca))
- Darth Vader ([@vader](https://github.com/vader))
- Princess Leia Organa ([@princess-leia](https://github.com/princess-leia))


## v5.0.0 (1980-05-17)

#### :boom: Breaking Change
* [#2](https://github.com/lerna/lerna-changelog/pull/2) chore: Terminate her... immediately! ([@gtarkin](https://github.com/gtarkin))

#### :bug: Bug Fix
* [#3](https://github.com/lerna/lerna-changelog/pull/3) fix: Get me the rebels base! ([@vader](https://github.com/vader))

#### Committers: 2
- Darth Vader ([@vader](https://github.com/vader))
- Governor Tarkin ([@gtarkin](https://github.com/gtarkin))


## v4.0.0 (1977-05-25)

#### :rocket: New Feature
* [#1](https://github.com/lerna/lerna-changelog/pull/1) feat: May the force be with you ([@luke](https://github.com/luke))

#### Committers: 1
- Luke Skywalker ([@luke](https://github.com/luke))"
`;

exports[`createMarkdown single tags outputs correct changelog 1`] = `
"
## Unreleased (2099-01-01)

#### :rocket: New Feature
* \`@star-wars/the-force-awakens\`, \`@star-wars/rogue-one\`
  * [#7](https://github.com/lerna/lerna-changelog/pull/7) feat: that is not how the Force works! ([@han-solo](https://github.com/han-solo))

#### :nail_care: Enhancement
* \`@star-wars/the-force-awakens\`, \`@star-wars/rogue-one\`
  * [#7](https://github.com/lerna/lerna-changelog/pull/7) feat: that is not how the Force works! ([@han-solo](https://github.com/han-solo))

#### Committers: 1
- Han Solo ([@han-solo](https://github.com/han-solo))


## v6.0.0 (1983-05-25)

#### :rocket: New Feature
* \`@star-wars/return-of-the-jedi\`
  * [#5](https://github.com/lerna/lerna-changelog/pull/5) feat: I am your father ([@vader](https://github.com/vader))

#### :bug: Bug Fix
* \`@star-wars/return-of-the-jedi\`
  * [#4](https://github.com/lerna/lerna-changelog/pull/4) fix: RRRAARRWHHGWWR ([@chewbacca](https://github.com/chewbacca))

#### :nail_care: Enhancement
* \`@star-wars/return-of-the-jedi\`
  * [#6](https://github.com/lerna/lerna-changelog/pull/6) refactor: he is my brother ([@princess-leia](https://github.com/princess-leia))

#### :house: Maintenance
* \`@star-wars/return-of-the-jedi\`
  * [#4](https://github.com/lerna/lerna-changelog/pull/4) fix: RRRAARRWHHGWWR ([@chewbacca](https://github.com/chewbacca))

#### Committers: 3
- Chwebacca ([@chewbacca](https://github.com/chewbacca))
- Darth Vader ([@vader](https://github.com/vader))
- Princess Leia Organa ([@princess-leia](https://github.com/princess-leia))


## v5.0.0 (1980-05-17)

#### :boom: Breaking Change
* \`@star-wars/empire-strikes-back\`
  * [#2](https://github.com/lerna/lerna-changelog/pull/2) chore: Terminate her... immediately! ([@gtarkin](https://github.com/gtarkin))

#### :bug: Bug Fix
* \`@star-wars/empire-strikes-back\`
  * [#3](https://github.com/lerna/lerna-changelog/pull/3) fix: Get me the rebels base! ([@vader](https://github.com/vader))

#### Committers: 2
- Darth Vader ([@vader](https://github.com/vader))
- Governor Tarkin ([@gtarkin](https://github.com/gtarkin))


## v4.0.0 (1977-05-25)

#### :rocket: New Feature
* \`@star-wars/a-new-hope\`
  * [#1](https://github.com/lerna/lerna-changelog/pull/1) feat: May the force be with you ([@luke](https://github.com/luke))

#### Committers: 1
- Luke Skywalker ([@luke](https://github.com/luke))"
`;
